lib/Native/NativeDecoder.php

```diff
@@ -0,0 +1,281 @@
+use LibDNS\Messages\Message;
+use LibDNS\Messages\MessageFactory;
+use LibDNS\Messages\MessageTypes;
+use LibDNS\Packets\PacketFactory;
+use LibDNS\Records\Question;
+use LibDNS\Records\QuestionFactory;
+use LibDNS\Records\Resource;
+use LibDNS\Records\Types\DomainName;
+use LibDNS\Records\Types\Type;
+use LibDNS\Records\Types\TypeBuilder;
+use LibDNS\Records\Types\Types;
+
+/**
+ * Decodes JSON DNS strings to Message objects.
+ *
+ * @author Daniil Gentili <https://daniil.it>, Chris Wright <https://github.com/DaveRandom>
+ */
+class NativeDecoder
+{
+    /**
+     * @var \LibDNS\Packets\PacketFactory
+     */
+    private $packetFactory;
+
+    /**
+     * @var \LibDNS\Messages\MessageFactory
+     */
+    private $messageFactory;
+
+    /**
+     * @var \LibDNS\Records\QuestionFactory
+     */
+    private $questionFactory;
+
+    /**
+     * @var \LibDNS\Records\Types\TypeBuilder
+     */
+    private $typeBuilder;
+    /**
+     * @var \LibDNS\Decoder\DecoderFactory
+     */
+    private $decoderFactory;
+
+    /**
+     * Map class names to IDs.
+     *
+     * @var array
+     */
+    private $classMap = [];
+    /**
+     * Constructor.
+     *
+     * @param \LibDNS\Packets\PacketFactory $packetFactory
+     * @param \LibDNS\Messages\MessageFactory $messageFactory
+     * @param \LibDNS\Records\QuestionFactory $questionFactory
+     * @param \LibDNS\Records\Types\TypeBuilder $typeBuilder
+     * @param \LibDNS\Encoder\EncodingContextFactory $encodingContextFactory
+     * @param \LibDNS\Decoder\DecoderFactory $decoderFactory
+     * @param bool $allowTrailingData
+     */
+    public function __construct(
+        PacketFactory $packetFactory,
+        MessageFactory $messageFactory,
+        QuestionFactory $questionFactory,
+        TypeBuilder $typeBuilder,
+        EncodingContextFactory $encodingContextFactory,
+        DecoderFactory $decoderFactory
+    ) {
+        $this->packetFactory = $packetFactory;
+        $this->messageFactory = $messageFactory;
+        $this->questionFactory = $questionFactory;
+        $this->typeBuilder = $typeBuilder;
+        $this->encodingContextFactory = $encodingContextFactory;
+        $this->decoderFactory = $decoderFactory;
+
+        $classes = new \ReflectionClass('\\LibDNS\\Records\\ResourceClasses');
+        foreach ($classes->getConstants() as $name => $value) {
+            $this->classMap[$name] = $value;
+        }
+    }
+    /**
+     * Decode a question record.
+     *
+     *
+     * @return \LibDNS\Records\Question
+     * @throws \UnexpectedValueException When the record is invalid
+     */
+    private function decodeQuestionRecord(string $name, int $type): Question
+    {
+        /** @var \LibDNS\Records\Types\DomainName $domainName */
+        $domainName = $this->typeBuilder->build(Types::DOMAIN_NAME);
+        $labels = \explode('.', $name);
+        if (!empty($last = \array_pop($labels))) {
+            $labels[] = $last;
+        }
```

Comment:
Looks like an infinite loop, or does it work, because there's a final dot? Also doesn't seem to work with domains like `0.a.c`.

Reply:
It's an if, not a while (you got me there for literally ten minutes).
And yes, it wouldn't work for domains like `a.c.0` (the opposite, shouldn't be an issue in most real-world usecases but fixed anyway)

Reply:
Oh...right. I guess the code should just be `$labels = \explode('.', \rtrim($labels, '.'))` then?